text/0000-dyno.md

```diff
@@ -0,0 +1,288 @@
+
+Note that `Requisition` has methods for providing values as well as references, and for providing more complex types. These will be covered in the next section.
+
+The important parts of `provide_any` are
+
+```rust
+pub mod provide_any {
+    pub trait Provider {
+        fn provide<'a>(&'a self, req: Requisition<'a, '_>);
+    }
+
+    pub fn request_by_type_tag<'a, I: TypeTag<'a>>(provider: &'a dyn Provider) -> Option<I::Type> { ... }
+
+    pub struct Requisition<'a, 'b>(...);
+
+    impl<'a, 'b> Requisition<'a, 'b> {
+        pub fn provide_ref<T: ?Sized + 'static>(&mut self, value: &'a T) -> &mut Self { ... }
+        ...
+    }
+
+    pub trait TypeTag<'a>: Sized + 'static {
+        type Type: 'a;
+    }
+
+    pub mod tags { ... }
+}
+```
+
+We have mostly covered how this interface is used above. The last piece is the concept of type tags, as used as a type parameter to `request_by_type_tag`.
+
+`core::any::TypeId` is a unique identifier for a type, but can only be used with types without lifetimes (or with only `'static` lifetimes). A type tag is an object which provides a layer of indirection to work around this limitation. For example, consider the type `&'l String`. It does not have a `TypeId` because of the `'l` lifetime. But we can use a type tag, `Ref<String>` to represent `&'l String` and `Ref<String>` does have a `TypeId`.
+
+The `TypeTag` trait is an abstraction over all type tags. It does not have any methods, only an associated type for the type which the tag represents. E.g., `<Ref<T> as TypeTag<'a>>::Type` is `&'a T`.
+
+There is no universal type tag. A concrete type tag must be written for a 'category' of types. A few common tags are provided in `provide_any::tags`, including `Value` for any type bounded by `'static`, and `Ref` for types of the form `&'a T` where `T: 'static`. For less common types, the user must provide a tag which implements `TypeTag`; in this way the `provide_any` API generalises to all types.
+
+Intermediate libraries (such as `error` in our examples) can choose to support only common type patterns so that the user is totally unaware of type tags, or a general API using type tags, or a hybrid approach where common cases are explicitly supported for ease of use and type tags are also supported for completeness.
+
+Note, that in the specialised variations, the user can use the type directly, e.g., `get_context_ref::<Backtrace>()>`. But in the general case the user must provide the type tag, not the type itself, e.g., `get_by_tag::<tags::Ref<Backtrace>>()`. That prevents using type inference or type ascription for access, and the type tag must always be explicitly specified.
```

Comment:
The trouble is this means that users have to know about type tags, which I find pretty unergonomic, in fact I consider it a primary advantage of the specialised versions that the user can name the types directly rather than using type tags. Note that, the user-facing API (the getters like get_context) always require an explicit type in any case, we can't rely on inference; it is only the library-facing API (provide_ref, etc) where the user can rely on inference.